### ISSUE TITLE
Useful extensions for the webgpuRenderer

### ISSUE THREAD

**Issue by Spiri0** [mrdoob/three.js#27274]
### Description

I plan to upload a v1 of my webgpu ocean to github after r159. This is a real simulation of the ocean according to physical laws, which means that the wave behavior is very realistic. This was only possible thanks to the many webgpu extensions in 3js, because this allowed me to use techniques that were previously only possible in unity and unreal. With webgpu this now also works in 3js. 
The lengthy and laborious error analyzes since June and the constant expansion of functionality in 3js thanks to Sunag and at times mrdoob have shown excellent results since r158.
But my claim is nothing less than a photorealistic ocean. The physicist in me comes through and just wants it to be perfect.
Physics is the language of nature.
That's why I'm already thinking about further necessary extensions for 3js.

To increase quality, StorageTextures would need the following:

mipMaps
linearMipmapLinearFilter
getMaxAnisotropy()

These mechanisms help to suppress unwanted noise effects at longer distances.
As a user, you should of course only create mipMaps if you no longer plan to do any further computing processes with the texture. So only at the end of all computing processes in each update interval before loading the generated textures into the final fragment shader. This is no different in unity.

When I tried to use FXAAShader I got an error message, but as a user I might be making a mistake and just have to do it differently with webgpu.

I couldn't see capabilities for getMaxAnisotropy() with Object.keys in the webgpuRenderer. Maybe this can be found somewhere else or hasn't been implemented yet? I'm also not aware of the possibility of using mipMaps with StorageTexture as well as with linearMipmapLinearFilter. Then I also got an error when i try to use it similar to the way in webgl

Photorealistic results require a lot of functionality but I don't see anything bad in that. This allows people to see what impressive possibilities 3js offers with webgpu. So far I've only been an enthusiastic user, but with my WebGPU project, which I want to use to create an ocean for my earth, I also want to make a contribution to the attractiveness of 3js

The green box only serves as a size reference (4m high and 2m deep in the water)
![fromnear](https://github.com/mrdoob/three.js/assets/69024222/39fece91-ca75-4b62-8f02-ad58152f2df5)
The foam behavior is also very real
![foamycrons](https://github.com/mrdoob/three.js/assets/69024222/31cc65e2-6274-42ed-aeb6-8fcccd0ef40a)

I will be able to close the gaps in the wire mesh again thanks to r159.

There is still a lot to do but the ocean project is one of my personal passions, alongside my real webgl landscape generator project with which I created the entire Earth and Mars (app size > 10GB because of the many maps). Over time I will also port this into a webgpu project as long as I can manage it alongside my job

### Solution

The possibility of using the following functionalities for StorageTexture:

mipMaps
linearMipmapLinearFilter
getMaxAnisotropy()

If the FXAAShader is not yet webgpuRenderer compatible, this too

### Alternatives

None, it's just about being able to use tried and tested functionalities from the webgl world in webgpu

**Comment by mrdoob** [comment]
That's looking really good!

**Comment by Spiri0** [comment]
Thank you, but there is much more possible. I'm currently trying to understand how to use the depthTexture in webgpu. The example in the examples is very nice, but when I pass the material with the depthTexture to the composer to check it, it doesn't work. But I'm still investigating, because the cause often lies with the user.
With a depthTexture I can then take objects under water and the ground into account with scattering and correct light refraction and absorption.

@Sunag
r159 works exactly as planned 👍I was able to reactivate my lod morphing without any texture distortions. The surface is now perfectly closed again and I can no longer see any transitions.

![smoothlodtransitions.PNG](https://github.com/mrdoob/three.js/assets/69024222/ac622159-32f9-43c6-a5eb-0c51bdc662b4)

I think I can upload a v1 to Github in December, even if a lot of things are still missing by my standards, such as: a GUI to change the waves. I know exactly where I have to adjust it to influence that and that's why I prefer to concentrate on understanding parts of the node system that I don't yet know how to use.
But then you can also experience what power the compute shaders have, which Sunag has now integrated very well. I use a lot of elements of the node system.

**Comment by sunag** [comment]
You have done an amazing project 👏 👏 

> I'm currently trying to understand how to use the depthTexture in webgpu. The example in the examples is very nice, but when I pass the material with the depthTexture to the composer to check it, it doesn't work

Could you provide a small example to reproduce the bug?

**Comment by Spiri0** [comment]
To test it, I simply added it directly into my threejs setup. It wouldn't stay there but the place is quite good for testing.
I don't get any error message. But the scene is darker and I suspect that I might have to clear something after I have rendered the depthTexture?

```
export const threejs_component = (() => {

    class ThreeJSController extends entity.Component {

        constructor() {
            super();
        }

        InitEntity() {

            if (WebGPU.isAvailable() === false) {
                document.body.appendChild(WebGPU.getErrorMessage());
                throw new Error('Your Browser does not support WebGPU yet');
            }

            this.threejs_ = new WebGPURenderer({
                canvas: document.createElement('canvas'),
                antialias: true,
            });

            this.threejs_.outputColorSpace = THREE.SRGBColorSpace;
            this.threejs_.setPixelRatio(window.devicePixelRatio);
            this.threejs_.shadowMap.enabled = true;
            this.threejs_.shadowMap.type = THREE.PCFSoftShadowMap;
            this.threejs_.physicallyCorrectLights = true;
            this.threejs_.domElement.id = 'threejs';

            this.container = document.getElementById('container');
            this.threejs_.setSize(this.container.clientWidth, this.container.clientHeight);
            this.container.appendChild( this.threejs_.domElement );

            const aspect = this.container.clientWidth / this.container.clientHeight;
            const fov = 50;
            const near = 0.1;
            const far = 1E6;
            this.camera_ = new THREE.PerspectiveCamera(fov, aspect, near, far);
            this.scene_ = new THREE.Scene();
            this.threejs_.setClearColor( 0x87CEEB );

            window.addEventListener('resize', () => {
                this.OnResize_();
            }, false);

 
            //---------------------------------------from example------------------------------------------
            this.depthTexture = new THREE.DepthTexture();
            this.depthTexture.type = THREE.FloatType;

            this.renderTarget = new THREE.RenderTarget(this.container.clientWidth, this.container.clientHeight);
            this.renderTarget.depthTexture = this.depthTexture;
            //-----------------------------------------------------------------------------------------------------
        }


        Render() {
            this.threejs_.setRenderTarget(this.renderTarget);
            this.threejs_.render(this.scene_, this.camera_);
            this.threejs_.setRenderTarget(null);

            this.camera_.layers.enableAll();
            this.threejs_.render(this.scene_, this.camera_);
        }

 
        Update(timeElapsed) {
        }


        OnResize_() {
            let width, height;

            if(window.innerWidth > window.innerHeight){
                width = 1.0 * window.innerWidth;
                height = 1.0 * window.innerHeight;              
            }      
            if(window.innerHeight > window.innerWidth){
                width = 1.0 * window.innerWidth;
                height = 1.0 * window.innerHeight;              
            }      
            this.camera_.aspect = width / height;
            this.camera_.updateProjectionMatrix();
            this.threejs_.setSize(width, height);  
        }
    }//end class

 
    return {
        ThreeJSController: ThreeJSController,
    };

})();
```

Later I can use the layer to determine what is rendered in the depthTexture. I found "_clearDepth" in the gpuRenderer but underscore commands are not intended for the user.
I think I need to clean up something after the depthTexture rendering but I don't know what yet.
Can I then pass the depthTexture to a wgslFN shader as normal?
I ask because when I tried this with composer I got error messages about the bindings.
I have now removed the composer because, as you mentioned, it is not part of the node system and probably only ran as usual by chance with the MeshBasicNodeMaterial. But I'm happy to reproduce this if you wish. But I prefer pure node solutions anyway and wait patiently from release to release. It's just a good concept that I want to use everywhere.

**Comment by Spiri0** [comment]
I've reduced this here, it was just too much unnecessary information. If I create a second camera "this.depthcamera_" the scene doesn't get darker.
Ok, I have no problems with that. I usually use an extra camera for depth textures anyway because I get more accurate results with a floating near for the depth camera.

But I think now I have something interesting for you. 
Error message in the console:
```
- While validating [ShaderModuleDescriptor "fragment"]

- While calling [Device].CreateShaderModule([ShaderModuleDescriptor "fragment"]).


Tint WGSL reader failure: :330:493 error: type mismatch for argument 30 in call to 'fragmentStage', expected 'texture_depth_2d', got 'f32'

```


The relevant pieces of code
```
const wgslShaderParams = {
   ...,
   depthTexture: texture(params.depthTexture),
}

this.oceanMaterial = new MeshBasicNodeMaterial();
this.oceanMaterial.colorNode = fragmentStageWGSL(wgslShaderParams);
```

```
fn fragmentStage(
   ...,
   depthTexture: texture_depth_2d,
)
```
I looked in the W3C docu. "texture_depth_2d" in chapter "16.7 Texture Build-in Functions" looks correct. I looked in nodes.js to see if there was an extra accessor for depthTextures, but I didn't find anything.



```
//depthTexture setup

        Init() {

            //threejs setups

            //----------------------------from WebGPU example---------------------------------------
            this.depthTexture = new THREE.DepthTexture();
            this.depthTexture.type = THREE.FloatType;

            this.renderTarget = new THREE.RenderTarget(this.container.clientWidth, this.container.clientHeight);
            this.renderTarget.depthTexture = this.depthTexture;
            //-----------------------------------------------------------------------------------------------------
        
        }

        Render() {
            this.threejs_.setRenderTarget(this.renderTarget);
            this.threejs_.render(this.scene_, this.depthcamera_);
            this.threejs_.setRenderTarget(null);

            this.camera_.layers.enableAll();
            this.threejs_.render(this.scene_, this.camera_);
        }

 
```

**Comment by Spiri0** [comment]
r160 promises to be exciting. Can the minFilter then be used with the usual three filters?

THREE.NearestFilter
THREE.LinearFilter
THREE.LinearMipmapLinearFilter

Together with the solved depthTexture issue and the maxAnisotropy, these are again major expansions of the webgpuRenderer.

I'm also looking forward to the post-processing nodes, but you can save that for later releases. I can well imagine that this could be a somewhat more extensive topic. And the patience has really paid off this year.

**Comment by sunag** [comment]
> r160 promises to be exciting. Can the minFilter then be used with the usual three filters?

With `sampler` this should work.

**Comment by Spiri0** [comment]
Hi Sunag, I'm currently investigating why artifacts occur when I use mipmaps normally. On this way:
```
var color = textureSample(texture, sampler, uv);
```
When I read and interpolate the mipmaps manually with "textureSampleLevel" everything is wonderful, it looks very nice. But it should also work in the normal way.

I tried to look into the sampler somehow but haven't found a way yet. I suspect something might be missing. Maybe e.g. the mipmap lod clamps, which means the shader may then process lods that don't exist.
The fine seams look suspiciously like antialiasing errors. Since everything works manually as it should and the mipmaps all look great, something must be going wrong during the interpolation. And the only thing that comes to mind is the sampler.
Can I configure the sampler manually instead of setting it with texture(myTexture)?
Then I could investigate further. I'm sure this will just be a small thing.

**Comment by sunag** [comment]
Hmm... maybe `sampler` node can work if you will use `wgslFn`. For example:

```js
const mySamplerTexture = new THREE.Texture(); // empty texture
mySamplerTexture... // sample settings

const myFn = wgslFn( `fn (  t: texture_2d<f32>, s: texture_type ) ...` );

const node = myFn( myTextureNode, mySamplerTexture )
```

**Comment by Spiri0** [comment]
That was a good idea from you, so I was able to do a second test with another sampler. 
Everything is fine with the samplers. So my hypothesis was wrong. But I was able to validate another one with the result.
The cause is the gradient formation in WGSL for the mipmap selection calculation. If I create the gradients with dpdx and dpdy and use their lengths to make the lod selection by myself, the exact same arthefact effect is created. The phenomenon would be very easy to solve if there was the “textureQueryLod” command in wgsl. Then I could simply read the lod directly next to the edge with a small delta from the edge and use the value for the respective edges as well. That would be easy and efficient. Without textureQueryLod, all mipmap lod logic has to be created manually in WGSL. I think WGSL still needs a little improvement in some areas.
Everything is fine on the 3js side. The lod selection via distance is not as efficient as with gradients but is stable at the edges because the position is always clear. That was proof for me yesterday that the mipmaps and wrapping are all correct.
If others would create issues about this topic in the future, you know that there are no errors. This is currently simply the nature of WGSL and requires manual lod selection. If wgsl were to introduce the command "textureQueryLod" or an analogue, then the solution as described would be quite simple for users.
